lib/services/expiry/expiry.go

```diff
@@ -0,0 +1,237 @@
+	// Log is the logger.
+	Log *slog.Logger
+	// Emitter is an events emitter, used to submit discrete events.
+	Emitter apievents.Emitter
+	// AccessPoint is a expiry access point.
+	AccessPoint authclient.ExpiryAccessPoint
+	// Clock is the clock used to calculate expiry.
+	Clock clockwork.Clock
+	// HostID is a unique ID of this host.
+	HostID string
+}
+
+// CheckAndSetDefaults checks required fields and sets default values.
+func (c *Config) CheckAndSetDefaults() error {
+	if c.Emitter == nil {
+		return trace.BadParameter("no Emitter configured for expiry")
+	}
+	if c.AccessPoint == nil {
+		return trace.BadParameter("no AccessPoint configured for expiry")
+	}
+	if c.Clock == nil {
+		c.Clock = clockwork.NewRealClock()
+	}
+	return nil
+}
+
+// Service is a expiry service.
+type Service struct {
+	*Config
+}
+
+// New initializes a expiry service
+func New(cfg *Config) (*Service, error) {
+	if err := cfg.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	s := &Service{
+		Config: cfg,
+	}
+	return s, nil
+}
+
+// Run starts the expiry service.
+func (s *Service) Run(ctx context.Context) error {
+	semCfg := services.SemaphoreLockConfigWithRetry{
+		SemaphoreLockConfig: services.SemaphoreLockConfig{
+			Service: s.AccessPoint,
+			Params: types.AcquireSemaphoreRequest{
+				SemaphoreKind: types.KindAccessRequest,
+				SemaphoreName: semaphoreName,
+				MaxLeases:     1,
+				Holder:        s.HostID,
+			},
+			Expiry: semaphoreExpiration,
+			Clock:  s.Clock,
+		},
+		Retry: retryutils.LinearConfig{
+			Clock:  s.Clock,
+			First:  time.Second,
+			Step:   semaphoreExpiration / 2,
+			Max:    semaphoreExpiration,
+			Jitter: retryutils.DefaultJitter,
+		},
+	}
+
+	poll := interval.New(interval.Config{
+		Duration:      scanInterval,
+		FirstDuration: retryutils.FullJitter(scanInterval),
+		Jitter:        retryutils.SeventhJitter,
+		Clock:         s.Clock,
+	})
+	defer poll.Stop()
+
+	for {
+		lease, err := services.AcquireSemaphoreLockWithRetry(ctx, semCfg)
+		if err != nil {
+			s.Log.WarnContext(ctx, "error acquiring semaphore", "error", err)
+			continue
+		}
+
+		if err := s.processRequests(ctx); err != nil {
+			s.Log.WarnContext(ctx, "error processing access requests", "error", err)
+		}
+		lease.Stop()
+		if err := lease.Wait(); err != nil {
+			s.Log.WarnContext(ctx, "error cleaning up semaphore", "error", err)
+		}
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-poll.Next():
+		}
+	}
+}
+func (s *Service) processRequests(ctx context.Context) error {
+	requestsExpired := 0
+	nextPageStart := ""
+	for {
+		var page []*types.AccessRequestV3
+		var err error
+		// Use time at read when calculating expiry of requests.
+		readTime := s.Clock.Now()
+		page, nextPageStart, err = s.getNextPageOfAccessRequests(ctx, nextPageStart)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		if len(page) == 0 {
+			return nil
+		}
+		minPageDelay := time.After(retryutils.SeventhJitter(minPageDelay))
```

Comment:
Don't think this is needed if you take my suggestion below.
```suggestion
```

Reply:
The reason for the minPageDelay variable being declared there is that we want it to start counting then, as it is the minimum time between reading the pages. Do you think it makes more sense to have a minimum delay between reading + processing instead of just the reading?

Reply:
@r0mant  Have moved the minPageDelay to where you suggested.

Reply:
If that change looks good to you I can go ahead and merge this

Reply:
@EdwardDowling Yeah go ahead and merge.